--- linux-2.6.24/arch/x86/boot/boot_old.h	2010-08-31 15:13:44.000000000 +0100
+++ linux-2.6.24/arch/x86/boot/boot.h	2010-08-31 15:09:15.000000000 +0100
@@ -109,7 +109,7 @@ typedef unsigned int addr_t;
 static inline u8 rdfs8(addr_t addr)
 {
 	u8 v;
-	asm volatile("movb %%fs:%1,%0" : "=r" (v) : "m" (*(u8 *)addr));
+	asm volatile("movb %%fs:%1,%0" : "=q" (v) : "m" (*(u8 *)addr));
 	return v;
 }
 static inline u16 rdfs16(addr_t addr)
